payu/manifest.py

```diff
@@ -47,87 +47,71 @@ def __init__(self, path,
         if ignore is not None:
             self.ignore = ignore
 
-        self.needsync = False
-        self.existing_filepaths = set()
-
-    def check_fast(self, reproduce=False, **args):
+    def calculate_fast(self, previous_manifest):
         """
-        Check hash value for all filepaths using a fast hash function and fall
-        back to slower full hash functions if fast hashes fail to agree.
+        Calculate hash value for all filepaths using a fast hash function and
+        fall back to slower full hash functions if fast hashes fail to agree,
+        with the pre-existing manifest
         """
-        hashvals = {}
-
-        fast_check = self.check_file(
+        # Calculate all fast hashes
+        self.add(
             filepaths=self.data.keys(),
-            hashvals=hashvals,
             hashfn=self.fast_hashes,
-            shortcircuit=True,
-            **args
+            force=True,
+            fullpaths=[self.fullpath(fpath) for fpath
+                       in list(self.data.keys())]
         )
 
-        if not fast_check:
-
-            # Save all the fast hashes for failed files that we've already
-            # calculated
-            for filepath in hashvals:
-                for hash, val in hashvals[filepath].items():
-                    self.data[filepath]['hashes'][hash] = val
-
-            if reproduce:
-                for filepath in hashvals:
-                    print('Check failed for {0} {1}'
-                          ''.format(filepath, hashvals[filepath]))
-                    tmphash = {}
-                    full_check = self.check_file(
-                        filepaths=filepath,
-                        hashfn=self.full_hashes,
-                        hashvals=tmphash,
-                        shortcircuit=False,
-                        **args
+        # If fast hashes from previous manifest match, use previous full hashes
+        # to avoid re-calculating slow hashes
+        self.update_matching_hashes(other=previous_manifest)
+
+        # Search for new files and files with changed fast hashes
+        changed_filepaths = set()
+        for filepath in self.data.keys():
+            for hash in self.data[filepath]['hashes'].values():
+                if hash is None:
+                    changed_filepaths.add(filepath)
+
+        # Calculate full hashes for these changed filepaths
+        if len(changed_filepaths) > 0:
+            self.add(
+                filepaths=list(changed_filepaths),
+                hashfn=self.full_hashes,
+                force=True,
+                fullpaths=[self.fullpath(fpath) for fpath
+                           in list(changed_filepaths)]
+            )
+
+    def check_reproduce(self, previous_manifest):
+        """
+        Compare full hashes with previous manifest
+        """
+        # Use paths in both manifests to pick up new and missing files
```

Comment:
yamanifest returns `None` for the hash value if the `filepath` is missing. By creating a set of all filepaths for previous and current manifests the missing files therefore just have `None` in the hash value. Am I correct?

As a result this isn't making a special case of missing files, i.e. they're not listed separately. Should they be?

Reply:
I lied above. This is also a question.

Reply:
Good point, it isn't making a special case. I can change the printed manifest errors to specify if it's a new or missing filepath - rather than `None != calculate_hash` or `hash != None` to make it more clear

Reply:
Added in  8fae6f9